pkg/controllers/pod/pod_controller.go

```diff
@@ -0,0 +1,328 @@
+func (r *PodReconciler) GarbageCollector(cancel chan bool, timeout time.Duration) {
+	ctx := context.Background()
+	log.Info("pod garbage collector started")
+	for {
+		select {
+		case <-cancel:
+			return
+		case <-time.After(timeout):
+		}
+		nsxSubnetPortSet := r.Service.ListNSXSubnetPortIDForPod()
+		if len(nsxSubnetPortSet) == 0 {
+			continue
+		}
+		podList := &v1.PodList{}
+		err := r.Client.List(ctx, podList)
+		if err != nil {
+			log.Error(err, "failed to list Pod")
+			continue
+		}
+
+		PodSet := sets.NewString()
+		for _, pod := range podList.Items {
+			PodSet.Insert(string(pod.UID))
+		}
+
+		for elem := range nsxSubnetPortSet {
+			if PodSet.Has(elem) {
+				continue
+			}
+			log.V(1).Info("GC collected Pod", "UID", elem)
+			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, MetricResTypePod)
+			err = r.Service.DeleteSubnetPort(types.UID(elem))
+			if err != nil {
+				metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResTypePod)
+			} else {
+				metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResTypePod)
+			}
+		}
+	}
+}
+
+func updateFail(r *PodReconciler, c *context.Context, o *v1.Pod, e *error) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, MetricResTypePod)
+}
+
+func deleteFail(r *PodReconciler, c *context.Context, o *v1.Pod, e *error) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResTypePod)
+}
+
+func updateSuccess(r *PodReconciler, c *context.Context, o *v1.Pod) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateSuccessTotal, MetricResTypePod)
+}
+
+func deleteSuccess(r *PodReconciler, _ *context.Context, _ *v1.Pod) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResTypePod)
+}
+
+func (r *PodReconciler) GetSubnetPathForPod(pod *v1.Pod) (string, error) {
+	subnetPath := ""
+	subnetSet, err := r.GetDefaultSubnetSet(pod)
+	if err != nil {
+		return "", err
+	}
+	log.Info("got default subnetset for pod", "subnetSet.UID", subnetSet.UID, "pod.Name", pod.Name)
+	subnetPath, err = r.AllocateSubnetFromSubnetSet(pod, subnetSet)
+	if err != nil {
+		return subnetPath, err
+	}
+	log.Info("allocated NSX subnet for pod", "nsxSubnetPath", subnetPath, "pod.Name", pod.Name)
+	return subnetPath, nil
+}
+
+// TODO: merge common functions for subnetport_controller and pod_controller
+func (r *PodReconciler) AllocateSubnetFromSubnetSet(pod *v1.Pod, subnetSet *v1alpha1.SubnetSet) (string, error) {
+	log.Info("allocating Subnet for Pod", "pod", pod.Name, "subnetSet", subnetSet.Name)
+	subnetPath, err := common.ServiceMediator.GetAvailableSubnet(subnetSet)
+	if err != nil {
+		log.Error(err, "failed to allocate Subnet")
+		return "", err
+	}
+	log.Info("allocated Subnet for Pod", "subnetPath", subnetPath)
+	return subnetPath, nil
+}
+
+func (r *PodReconciler) GetDefaultSubnetSet(pod *v1.Pod) (*v1alpha1.SubnetSet, error) {
```

Comment:
Do we need to specify the purpose of this func? GetDefaultPodSubnetSet?

Reply:
I polished the code and now there's a public function `GetDefaultSubnetSet` in `pkg/controllers/common/utils.go` which can be invoked by both subnetport_controller and pod_controller